### ISSUE TITLE
which optimize option use in Scala 2.12

### ISSUE THREAD

**Issue by xuwei-k** [scalaz/scalaz#1292]
http://scala-lang.org/news/2.12.0#new-optimizer

```
$ scala -version
Scala code runner version 2.12.0 -- Copyright 2002-2016, LAMP/EPFL and Lightbend, Inc.
$ scala -opt:help
Enable optimizations
  unreachable-code     Eliminate unreachable code, exception handlers guarding no instructions, redundant metadata (debug information, line numbers).
  simplify-jumps       Simplify branching instructions, eliminate unnecessary ones.
  compact-locals       Eliminate empty slots in the sequence of local variables.
  copy-propagation     Eliminate redundant local variables and unused values (including closures). Enables unreachable-code.
  redundant-casts      Eliminate redundant casts using a type propagation analysis.
  box-unbox            Eliminate box-unbox pairs within the same method (also tuples, xRefs, value class instances). Enables unreachable-code.
  nullness-tracking    Track nullness / non-nullness of local variables and apply optimizations.
  closure-invocations  Rewrite closure invocations to the implementation method.
  inline-project       Inline only methods defined in the files being compiled. Enables unreachable-code.
  inline-global        Inline methods from any source, including classfiles on the compile classpath. Enables unreachable-code.
  l:none               Disable optimizations. Takes precedence: `-opt:l:none,+box-unbox` / `-opt:l:none -opt:box-unbox` don't enable box-unbox.
  l:default            Enable default optimizations: unreachable-code.
  l:method             Enable intra-method optimizations: unreachable-code,simplify-jumps,compact-locals,copy-propagation,redundant-casts,box-unbox,nullness-tracking,closure-invocations.
  l:project            Enable cross-method optimizations within the current project: l:method,inline-project.
  l:classpath          Enable cross-method optimizations across the entire classpath: l:project,inline-global.

**Comment by S11001001** [comment]
Hmm, `l:method` seems like a winner.  Only one I'm not sure about there is `compact-locals`.

Until scala.meta arrives and gives us proper inlining (and a proper way to offer it to users), not sure that's a hole worth going down.  If we want to try anyway, `l:project` and a sprinkling of `@inline` in "obvious" places (to be replaced with meta's `inline` keyword later).

**Comment by S11001001** [comment]
Merged #1338.
